Infinite scroll

KarlOlofA opened the pull request.

KarlOlofA committed:
Infinite Scroll & Filter Functionality

Added functionality for scrolling to infinity!
Also a filter functionality between Top & New stories.

KarlOlofA committed:
Fixed error regarding variables with identical names

KarlOlofA committed:
Another naming error

ostseb commented on src/app/fetch-stories.tsx:
```diff
@@ -0,0 +1,37 @@
+"use server";
```
This file could perhaps be merged with src/services/hn.ts

KarlOlofA replied:
Moved fetching to hn.ts

ostseb commented on src/app/fetch-stories.tsx:
```diff
@@ -0,0 +1,37 @@
+    const start = fetchData.page*perPage;
+    const end = start + perPage;
+    const storiesToFetch = stories.slice(start, end);
+    const hasMoreStories = storiesToFetch.length > 0;
```
```suggestion
    const hasMoreStories = storiesToFetch.length == perPage;
```

Changing it to this would detect the last page earlier, if ex the last page is only returning 5 items.

KarlOlofA replied:
Fixed! :D

ostseb commented on src/app/fetch-stories.tsx:
```diff
@@ -0,0 +1,37 @@
+    const storiesToFetch = stories.slice(start, end);
+    const hasMoreStories = storiesToFetch.length > 0;
+
+    const fetchedStories= await Promise.all(storiesToFetch.map(id => fetch(`https://hacker-news.firebaseio.com/v0/item/${id}.json`, {
```
Great that we slice before we fetch all the details so we only fetch the data we need for the current page. 👍

KarlOlofA replied:
💯 🥳

ostseb reviewed (commented):
Not sure what the issue is, but your infinite scroll is only working after I’ve toggled new/top stories once, does not work before. I’m also missing is some sort of persistent state of the current page. Now going back from an article page we always start over on the first page.  For the first time in Next/React it looks good! Great job!

KarlOlofA committed:
Bug fixing & Restructuring

The page now infinately scrolls on first load. Was a problem with the last renaming pushes that solved some other errors.
Moved fetching code to hn.ts.

KarlOlofA committed:
Changed a small check that handles if we should load more stories

Changed hasMoreStories check.
